crates/turborepo-lib/src/daemon/server.rs

```diff
@@ -469,60 +468,50 @@ impl proto::turbod_server::Turbod for TurboGrpcServiceInner {
-                        })
-                        .collect(),
-                    package_manager: proto::PackageManager::from(packages.package_manager).into(),
-                })
-            })
-            .map_err(|e| match e {
-                turborepo_repository::discovery::Error::Unavailable => {
-                    tonic::Status::unavailable("package discovery unavailable")
-                }
-                turborepo_repository::discovery::Error::Failed(e) => {
-                    tonic::Status::internal(format!("{}", e))
-                }
-            })
+        match self.package_watcher.discover_packages().await {
+            Some(Ok(packages)) => Ok(tonic::Response::new(proto::DiscoverPackagesResponse {
+                package_files: packages
+                    .workspaces
+                    .into_iter()
+                    .map(|d| proto::PackageFiles {
+                        package_json: d.package_json.to_string(),
+                        turbo_json: d.turbo_json.map(|t| t.to_string()),
+                    })
+                    .collect(),
+                package_manager: proto::PackageManager::from(packages.package_manager).into(),
+            })),
+            None | Some(Err(PackageWatchError::Unavailable)) => {
+                Err(tonic::Status::unavailable("package discovery unavailable"))
+            }
+            Some(Err(PackageWatchError::InvalidState(reason))) => {
+                Err(tonic::Status::failed_precondition(reason))
+            }
+        }
     }
 
     async fn discover_packages_blocking(
         &self,
         _request: tonic::Request<proto::DiscoverPackagesRequest>,
     ) -> Result<tonic::Response<proto::DiscoverPackagesResponse>, tonic::Status> {
-        self.package_discovery
-            .discover_packages_blocking()
-            .await
-            .map(|packages| {
-                tonic::Response::new(proto::DiscoverPackagesResponse {
-                    package_files: packages
-                        .workspaces
-                        .into_iter()
-                        .map(|d| proto::PackageFiles {
-                            package_json: d.package_json.to_string(),
-                            turbo_json: d.turbo_json.map(|t| t.to_string()),
-                        })
-                        .collect(),
-                    package_manager: proto::PackageManager::from(packages.package_manager).into(),
-                })
-            })
-            .map_err(|e| match e {
-                turborepo_repository::discovery::Error::Unavailable => {
-                    tonic::Status::unavailable("package discovery unavailable")
-                }
-                turborepo_repository::discovery::Error::Failed(e) => {
-                    tonic::Status::internal(format!("{}", e))
-                }
-            })
+        match self.package_watcher.discover_packages_blocking().await {
```

Comment:
It blocks until there's a non-None value available. Not necessarily a success. When we detect relevant changes, we set the value to `None`, then recalculate. That recalculation might produce an error, in which case this would resolve to an error.

Reply:
I believe the blocking variant is used by the LSP, while the non-blocking variant is for the CLI, which will fall back to running its own calculation in the event of a timeout